File: apps/backend/src/routes/api/v1/folders.ts
import { Hono } from 'hono';
import { zValidator } from '@hono/zod-validator';
import { folderService } from '../../../services/folder.service';
import { authMiddleware } from '../../../middleware/auth';
import { createFolderSchema } from '../../../utils/validators-extended';

const folderRouter = new Hono();

folderRouter.post('/', authMiddleware, zValidator('json', createFolderSchema), async (c) => {
  const user = c.get('user');
  const { name, parentId } = c.req.valid('json');

  const folder = await folderService.createFolder(user.id, name, parentId);
  return c.json(folder);
});

folderRouter.get('/', authMiddleware, async (c) => {
  const user = c.get('user');
  const folders = await folderService.getFolders(user.id);
  return c.json(folders);
});

folderRouter.get('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  const folder = await folderService.getFolderById(id, user.id);
  return c.json(folder);
});

folderRouter.put('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');
  const data = await c.req.json();

  const folder = await folderService.updateFolder(id, user.id, data);
  return c.json(folder);
});

folderRouter.delete('/:id', authMiddleware, async (c) => {
  const user = c.get('user');
  const id = c.req.param('id');

  await folderService.deleteFolder(id, user.id);
  return c.json({ message: 'Folder deleted' });
});

export default folderRouter;

File: apps/backend-python/src/routes/api/v1/folders.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from typing import Optional
from sqlalchemy.orm import Session
from src.config.database import get_db
from src.middleware.auth import get_current_user
from src.services.folder_service import folder_service
from src.db.models import User

router = APIRouter()


class CreateFolderSchema(BaseModel):
    name: str
    parentId: Optional[str] = None


class UpdateFolderSchema(BaseModel):
    name: Optional[str] = None
    parentId: Optional[str] = None
    isExpanded: Optional[bool] = None


@router.post("")
def create_folder(data: CreateFolderSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return folder_service.create_folder(db, user.id, data.name, data.parentId)


@router.get("")
def list_folders(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return folder_service.get_folders(db, user.id)


@router.get("/{folder_id}")
def get_folder(folder_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return folder_service.get_folder_by_id(db, folder_id, user.id)


@router.put("/{folder_id}")
def update_folder(folder_id: str, data: UpdateFolderSchema, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    return folder_service.update_folder(db, folder_id, user.id, data.model_dump(exclude_none=True))


@router.delete("/{folder_id}")
def delete_folder(folder_id: str, user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    folder_service.delete_folder(db, folder_id, user.id)
    return {"message": "Folder deleted"}
